subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <string.h>
+#include <errno.h>
+#include <cache.h>
+#include <ipc/ipc_icmsg_buf.h>
+
+
+/* Helpers */
+static uint32_t idx_occupied(uint32_t len, uint32_t a, uint32_t b)
+{
+	/* It is implicitly assumed a and b cannot differ by more then len. */
+	return (b > a) ? (len - (b - a)) : (a - b);
+}
+
+static uint32_t idx_cut(uint32_t len, uint32_t idx)
+{
+	/* It is implicitly assumed a and b cannot differ by more then len. */
+	return (idx >= len) ? (idx - len) : (idx);
+}
+
+struct icmsg_buf *icmsg_buf_init(void *buf, size_t blen)
+{
+	/* blen must be big enough to contain icmsg_buf struct, byte of data
+	 * and message len (2 bytes).
+	 */
+	struct icmsg_buf *ib = buf;
+
+	__ASSERT_NO_MSG(blen > (sizeof(*ib) + sizeof(uint16_t)));
+
+	ib->len = blen - sizeof(*ib);
+	ib->wr_idx = 0;
+	ib->rd_idx = 0;
+
+	__sync_synchronize();
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_WB);
+
+	return ib;
+}
+
+int icmsg_buf_write(struct icmsg_buf *ib, const char *buf, uint16_t len)
+{
+	/* The length of buffer is immutable - avoid reloading that may happen
+	 * due to memory bariers.
+	 */
+	const uint32_t iblen = ib->len;
+
+	/* rx_idx == wr_idx means the buffer is empty.
+	 * Max bytes that can be stored is len - 1.
+	 */
+	const uint32_t max_len = iblen - 1;
+
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint32_t wr_idx = ib->wr_idx;
+	uint32_t rd_idx = ib->rd_idx;
+
+	if ((len + sizeof(len) > max_len) || (len == 0)) {
+		/* Incorrect call. */
+		return -EINVAL;
+	}
+
+	uint32_t avail = max_len - idx_occupied(iblen, wr_idx, rd_idx);
+
+	if (len + sizeof(len) > avail) {
+		/* No free space. */
+		return -ENOMEM;
+	}
+
+	/* Store info about the message length. */
+	ib->data[wr_idx] = (uint8_t)len;
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+	ib->data[wr_idx] = (uint8_t)(len >> 8);
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
```

Comment:
Flush cache holding parts of len
`sys_cache_data_range(&ib->data[wr_idx], sz, K_CACHE_WB);` here and after line 78